### PR TITLE
Stop using pickle, add some tests

### PR DESCRIPTION
**Pull request by jamesandres**

### CODE REVIEW

File: requirements.pip [hunk 1]
```diff
@@ -1 +1 @@
-Django>=1.2
+Django>=1.2,<=1.4.14
```
Comment (lanzz) [inline requirements.pip]: Why not just `==1.4.14`?

Comment (jamesandres) [reply requirements.pip]: Wellll, this project was a fork of django-dirtyfields circa 2012 and we are trying to not change things unnecessarily

File: src/dirtyfields/dirtyfields.py [hunk 2]
```diff
@@ -37,13 +35,12 @@ def _get_field_value(self, f):
         # Else use field name, as can actually be set to be different from db column name
         else:
             val = getattr(self, f.name)
-        if f.column in self._pickle_fields:
-            val = pickle.dumps(val)
+
+        if not isinstance(val, IMMUTABLE_TYPES):
+            val = copy(val)
```
Comment (lanzz) [inline src/dirtyfields/dirtyfields.py]: Is it safe to do a shallow copy here? What if something has changed deeper than first level?

File: src/dirtyfields/dirtyfields.py [hunk 3]
```diff
@@ -37,13 +35,12 @@ def _get_field_value(self, f):
...
         return val
 
-    def _get_value(self, val, col_name, unpickle):
-        if unpickle and col_name in self._pickle_fields:
-            val = pickle.loads(val)
+    def _get_value(self, val, col_name):
```
Comment (lanzz) [inline src/dirtyfields/dirtyfields.py]: Retire this method altogether?

File: tests/django_settings.py [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+SECRET_KEY = 'WE DONT CARE ABOUT IT'
+
+DATABASES = {
+    'default': {
+        'ENGINE': 'django.db.backends.sqlite3', # Add 'postgresql_psycopg2', 'postgresql', 'mysql', 'sqlite3' or 'oracle'.
+        'NAME': 'test.db',                      # Or path to database file if using sqlite3.
+        'USER': '',                      # Not used with sqlite3.
+        'PASSWORD': '',                  # Not used with sqlite3.
+        'HOST': '',                      # Set to empty string for localhost. Not used with sqlite3.
+        'PORT': '',                      # Set to empty string for default. Not used with sqlite3.
```
Comment (lanzz) [inline tests/django_settings.py]: Not sure there's much point in keeping the comments here

File: tests/test_dirtyfields.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+    def test_makes_copy_of_mutable_field(self):
+        a = Animal(name='Duck',
+                   list_of_things=['wing', 'bill', 'webbed foot'])
+        self.assertNotEqual(
+            id(a.list_of_things),
+            id(a._original_state['list_of_things']))
+        a.list_of_things[0] = 'tail'
+        self.assertEqual(a.get_dirty_fields(),
+                         {'list_of_things': ['wing', 'bill', 'webbed foot']})
```
Comment (lanzz) [inline tests/test_dirtyfields.py]: Add another test similar to this, but for fields changed post-save

File: requirements.pip (hunk 1 above)
Comment (lanzz) [reply requirements.pip]: I think limiting it to the ancient 1.4.14 is big enough change :) If _anyone_ is using this specific fork with versions other than 1.4.14, it's definitely going to be newer not older versions :)

File: src/dirtyfields/dirtyfields.py (hunk 2 above)
Comment (jamesandres) [reply src/dirtyfields/dirtyfields.py]: Excellent question, possibly _the best_ question. Go have a look at the other PR, we put a control in there to ensure people don't do this.

File: tests/django_settings.py (hunk 4 above)
Comment (jamesandres) [reply tests/django_settings.py]: I copied this from the upstream version of dirtyfields, can remove though

File: src/dirtyfields/dirtyfields.py (hunk 2 above)
Comment (lanzz) [reply src/dirtyfields/dirtyfields.py]: I've seen it, but it isn't strict enough, immutable containers can contain mutable items; you're guaranteed the _item_ will remain the same (you can't replace a list item inside a tuple with a different list), but not that its _contents_ will remain the same (you can still push stuff into the list)

File: tests/django_settings.py (hunk 4 above)
Comment (lanzz) [reply tests/django_settings.py]: Fair enough if it comes from upstream, let's keep it

File: src/dirtyfields/dirtyfields.py (hunk 2 above)
Comment (jamesandres) [reply src/dirtyfields/dirtyfields.py]: The point is, our fork of "django-dirtyfields" does not need to be bulletproof it only needs to support the conversocial codebase. We simply "won't do that".

The community version of django-dirtyfields is lightyears ahead of this code and does indeed very carefully solve these problems (eg: using deepcopy). But at a cost of performance.

File: tests/django_settings.py (hunk 4 above)
Comment (jamesandres) [reply tests/django_settings.py]: eehh we are forked so hard now we might as well fix it

File: requirements.pip (hunk 1 above)
Comment (jamesandres) [reply requirements.pip]: I don't care much either way, the only thing using this code is the conversocial monolith and we have django pinned at 1.4.14

File: src/dirtyfields/dirtyfields.py (hunk 2 above)
Comment (lanzz) [reply src/dirtyfields/dirtyfields.py]: I think this needs to be put explicitly into writing both here and in the monolith codebase

File: requirements.pip (hunk 1 above)
Comment (lanzz) [reply requirements.pip]: That was my thinking exactly when I suggested to pin it to `==1.4.14` :)

File: src/dirtyfields/dirtyfields.py (hunk 2 above)
Comment (jamesandres) [reply src/dirtyfields/dirtyfields.py]: I'll add a comment though to explain this decision better

**Comment by lanzz** [comment]
🐳

**Pull request merged**